File: Data Structures & Algorithms/combination-target-sum-ii/submission-3.py
class Solution:
    def combinationSum2(self, candidates: List[int], target: int) -> List[List[int]]:
        res = []
        candidates.sort()
            
        def helper(index, total, curr):
            if total == target:
                res.append(curr.copy())
                return
            elif total > target or index >= len(candidates):
                return

            num = candidates[index]
            i = index

            curr.append(num)
            helper(index + 1, total + num, curr)
            curr.pop()
            while i < len(candidates) and candidates[i] == num:
                i += 1
            if i >= len(candidates):
                return
            helper(i, total, curr)
        
        helper(0, 0, [])
        
        return res

File: Data Structures & Algorithms/target-sum/submission-3.java
class Solution {
    private int[] nums;
    private int target;

    private int res = 0;

    public int findTargetSumWays(int[] nums, int target) {
        this.nums = nums;
        this.target = target;

        helper(0, 0);
        return res;
    }

    private void helper(int index, int current) {
        if (index == nums.length) {
            if (current == target) res += 1;
            return;
        }
        
        helper(index + 1, current + nums[index]);
        helper(index + 1, current - nums[index]);
    }
}
